src/reanimated2/helperTypes.ts

```diff
@@ -22,102 +24,69 @@ import type { ReanimatedKeyframe } from './layoutReanimation/animationBuilder/Ke
- */
-export type TransformStyleTypes = ExtractArrayItemType<RNTransformType>;
-
-export type TransformArrayItemType = TransformStyleTypes;
-
-// Note: why is `readonly` here? For safety.
-// In TS `ReadonlyArray` is a supertype of `Array`,
-// therefore `Array` can be assigned to `ReadonlyArray` but
-// `ReadonlyArray` cannot be assigned to `Array`.
-export type AnimatedTransform =
-  | readonly (TransformArrayItemType | SharedValue<TransformArrayItemType>)[]
-  | RNTransformType;
-
-/**
- * @deprecated Please use `AnimatedStyle` type instead.
- */
-export type AnimateStyle<S> = {
-  [K in keyof S]: K extends 'transform'
-    ? AnimatedTransform
-    : S[K] extends ReadonlyArray<any>
-    ? ReadonlyArray<AnimateStyle<S[K][0]>>
-    : S[K] extends object
-    ? AnimateStyle<S[K]>
-    : S[K] extends ColorValue | undefined
-    ? S[K] | number
-    : S[K] | SharedValue<AnimatableValue>;
-};
-
-export type AnimatedStyle<S> = AnimateStyle<S>;
-
-// provided types can either be their original types (like backgroundColor: pink)
-// or inline shared values/derived values
-type MaybeSharedValue<S> = {
-  [K in keyof S]: S[K] | Readonly<SharedValue<Extract<S[K], AnimatableValue>>>;
-};
-
-export type StylesOrDefault<T> = 'style' extends keyof T
-  ? MaybeSharedValue<T['style']>
-  : Record<string, unknown>;
+type MaybeSharedValueRecursive<TValue> = TValue extends (infer TItem)[]
+  ? SharedValue<TItem[]> | (MaybeSharedValueRecursive<TItem> | TItem)[]
+  : TValue extends object
+  ?
+      | SharedValue<TValue>
+      | {
+          [TKey in keyof TValue]:
+            | MaybeSharedValueRecursive<TValue[TKey]>
+            | TValue[TKey];
+        }
+  : MaybeSharedValue<TValue>;
+
+type DefaultStyle = ViewStyle & ImageStyle & TextStyle;
+
+// Ideally we want AnimatedStyle to not be generic, but there are
+// so many depenedencies on it being generic that it's not feasible at the moment.
+export type AnimatedStyle<Style = DefaultStyle> =
+  | Style
+  | MaybeSharedValueRecursive<Style>;
 
 type EntryOrExitLayoutType =
   | BaseAnimationBuilder
   | typeof BaseAnimationBuilder
   | EntryExitAnimationFunction
   | ReanimatedKeyframe;
 
-/* 
+/*
   Style type properties (properties that extends StyleProp<ViewStyle>)
   can be defined with other property names than "style". For example `contentContainerStyle` in FlatList.
   Type definition for all style type properties should act similarly, hence we
   pick keys with 'Style' substring with the use of this utility type.
 */
-type PickStyleProps<T> = Pick<
-  T,
+type PickStyleProps<TProps> = Pick<
+  TProps,
   {
-    [Key in keyof T]-?: Key extends `${string}Style` ? Key : never;
-  }[keyof T]
+    [Key in keyof TProps]-?: Key extends `${string}Style` | 'style'
+      ? Key
+      : never;
+  }[keyof TProps]
 >;
 
-type StyleAnimatedProps<P extends object> = {
-  [K in keyof PickStyleProps<P>]: StyleProp<
-    AnimatedStyle<P[K] | MaybeSharedValue<P[K]>>
+type StyleProps<TProps extends object> = {
```

Comment:
IDK what is the better name, however these styles are animated, so maybe:
```suggestion
type AnimatedStyleProps<TProps extends object> = {
```